bn-apps/cordapp-updates-distribution/corda-updates-core/src/main/kotlin/net/corda/businessnetworks/cordaupdates/core/CordappSyncer.kt

```diff
@@ -0,0 +1,122 @@
+     */
+    fun syncCordapps(coordinatesWithoutVersion : String? = null, additionalConfigurationProperties : Map<String, Any> = mapOf()) : List<ArtifactMetadata> {
+        return if (coordinatesWithoutVersion == null) {
+            syncerConf.cordappSources.flatMap { syncTask ->
+                val resolver = CordaMavenResolver.create(syncerConf, syncTask, repositoryListener = repositoryListener, transferListener = transferListener)
+                syncTask.cordapps.map {
+                    resolver.downloadVersionRange("$it:[,)", additionalConfigurationProperties)
+                }
+            }
+        } else {
+            val source = syncerConf.findSourceFor(coordinatesWithoutVersion) ?: throw CordappSourceNotFoundException(coordinatesWithoutVersion)
+            val resolver = CordaMavenResolver.create(syncerConf, source)
+            listOf(resolver.downloadVersionRange(coordinatesWithoutVersion, additionalConfigurationProperties))
+        }
+    }
+
+    /**
+     * Returns a list of CorDapp versions available in the remote repository. syncerConf is expected to contain a configured [CordappSource] for the requested CorDapp.
+     *
+     * @param coordinatesWithRange full maven coordinates of a CorDapp with a version range, i.e. "net.corda:corda-finance:[0,2.5)".
+     * @param additionalConfigurationProperties additional parameters to be passed to [CordaMavenResolver]
+     */
+    fun getAvailableVersions(coordinatesWithRange : String, additionalConfigurationProperties : Map<String, Any> = mapOf()) : List<ArtifactMetadata> {
+        val syncTask = syncerConf.findSourceFor(coordinatesWithRange) ?: throw CordappSourceNotFoundException(coordinatesWithRange)
+        val resolver = CordaMavenResolver.create(syncerConf, syncTask)
+        return listOf(resolver.resolveVersionRange(coordinatesWithRange, additionalConfigurationProperties))
+    }
+}
+
+/**
+ * [CordappSource] allows, that allows to associate multiple CorDapps with a remote repository
+ */
+data class CordappSource(
+        val remoteRepoUrl : String,
+        val cordapps : List<String>,
+        val httpUsername : String? = null,
+        val httpPassword : String? = null)
+
+/**
+ * Configuration for [CordappSyncer]. Usually is red from settings.conf
+ */
+data class SyncerConfiguration(
+        val localRepoPath : String,
+        val httpProxyHost : String? = null,
+        val httpProxyType : String? = null,
+        val httpProxyPort : Int? = null,
+        val httpProxyUsername : String? = null,
+        val httpProxyPassword : String? = null,
+        val rpcHost : String? = null,
+        val rpcPort : Int? = null,
+        val rpcUsername : String? = null,
+        val rpcPassword : String? = null,
+        val cordappSources : List<CordappSource>) {
+    companion object {
+        fun readFromFile(file : File) : SyncerConfiguration {
+
+            val conf = ConfigFactory.parseFile(file)!!
```

Comment:
Can't Typesafe Config map the fields to this object directly? Why do you have to copy every field here?

Reply:
It can, but then objects have to follow java beans convention, i.e. no data classes, mutable fields, zero-args constructor, manual `equals` and `hashCode` and etc.